client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -633,10 +633,20 @@ public void closeAndFree(
                     log.debug("sender open [target: {}, sendQueueFull: {}]", targetAddress, sender.sendQueueFull());
                     // wait on credits a little time, if not already given
                     if (sender.getCredit() <= 0) {
-                        vertx.setTimer(clientConfigProperties.getFlowLatency(), timerID -> {
-                            log.debug("sender [target: {}] has {} credits after grace period of {}ms", targetAddress,
-                                    sender.getCredit(), clientConfigProperties.getFlowLatency());
-                            result.tryComplete(sender);
+                        final long waitOnCreditsTimerId = vertx.setTimer(clientConfigProperties.getFlowLatency(),
+                                timerID -> {
+                                    log.debug("sender [target: {}] has {} credits after grace period of {}ms",
+                                            targetAddress,
+                                            sender.getCredit(), clientConfigProperties.getFlowLatency());
+                                    sender.sendQueueDrainHandler(null);
+                                    result.tryComplete(sender);
+                                });
+                        sender.sendQueueDrainHandler(replenishedSender -> {
+                            log.debug("sender [target: {}] has received {} initial credits",
+                                    targetAddress, replenishedSender.getCredit());
+                            vertx.cancelTimer(waitOnCreditsTimerId);
+                            replenishedSender.sendQueueDrainHandler(null);
+                            result.tryComplete(replenishedSender);
```

Comment:
IMHO it would be cleaner to actually check for the time having been cancelled:
```
if (vertx.cancelTimer(waitOnCreditsTimerId) {
  result.tryComplete(replenishedSender);
} // otherwise the timer has already completed the future
```

Reply:
OK.